src/descriptor/key.rs

```diff
@@ -717,6 +730,115 @@ impl MiniscriptKey for DescriptorPublicKey {
+        self.key.derive_public_key(self.secp)
+    }
+
+    /// Return the derivation index of this key
+    pub fn get_index(&self) -> u32 {
+        self.index
+    }
+
+    /// Construct an instance from a descriptor key and a derivation index
+    ///
+    /// Panics if the key contains a wildcard
+    fn new(key: DescriptorPublicKey, index: u32, secp: &'secp Secp256k1<C>) -> Self {
+        if let DescriptorPublicKey::XPub(ref xpk) = &key {
+            assert!(
+                xpk.wildcard == Wildcard::None,
+                "Derived descriptor keys cannot contain any wildcards"
+            );
+        }
+
+        DerivedDescriptorKey { key, index, secp }
+    }
+}
+
+impl<'secp, C: Verification> fmt::Display for DerivedDescriptorKey<'secp, C> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        self.key.fmt(f)
+    }
+}
+impl<'secp, C: Verification> fmt::Debug for DerivedDescriptorKey<'secp, C> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.debug_struct("DerivedDescriptorKey")
+            .field("key", &self.key)
+            .field("index", &self.index)
+            .finish()
+    }
+}
+
+impl<'secp, C: Verification> PartialEq for DerivedDescriptorKey<'secp, C> {
+    fn eq(&self, other: &Self) -> bool {
+        self.key.eq(&other.key) && self.index.eq(&other.index)
+    }
+}
+impl<'secp, C: Verification> Eq for DerivedDescriptorKey<'secp, C> {}
+
+impl<'secp, C: Verification> PartialOrd for DerivedDescriptorKey<'secp, C> {
+    fn partial_cmp(&self, other: &Self) -> Option<cmp::Ordering> {
+        Some(self.cmp(other))
+    }
+}
+impl<'secp, C: Verification> Ord for DerivedDescriptorKey<'secp, C> {
+    fn cmp(&self, other: &Self) -> cmp::Ordering {
+        self.key
+            .cmp(&other.key)
+            .then_with(|| self.index.cmp(&other.index))
+    }
+}
+
+impl<'secp, C: Verification> Clone for DerivedDescriptorKey<'secp, C> {
+    fn clone(&self) -> Self {
+        DerivedDescriptorKey {
+            key: self.key.clone(),
+            index: self.index,
+            secp: self.secp,
+        }
+    }
+}
+
+impl<'secp, C: Verification> hash::Hash for DerivedDescriptorKey<'secp, C> {
+    fn hash<H: hash::Hasher>(&self, state: &mut H) {
+        self.key.hash(state);
+        self.index.hash(state);
+    }
+}
+
+impl<'secp, C: Verification> MiniscriptKey for DerivedDescriptorKey<'secp, C> {
+    // This allows us to be able to derive public keys even for PkH s
+    type Hash = Self;
+
+    fn is_uncompressed(&self) -> bool {
+        self.key.is_uncompressed()
+    }
+
+    fn is_x_only_key(&self) -> bool {
+        self.key.is_x_only_key()
+    }
+
+    fn to_pubkeyhash(&self) -> Self {
+        self.clone()
+    }
+}
+
+impl<'secp, C: Verification> ToPublicKey for DerivedDescriptorKey<'secp, C> {
+    fn to_public_key(&self) -> bitcoin::PublicKey {
+        self.key.derive_public_key(&self.secp).unwrap()
```

Comment:
Let us not derive `ToPublicKey` at all. Users can also use the https://github.com/rust-bitcoin/rust-miniscript/blob/d8cc633b5b726e7ae595031ba04dcc1c1c2d6fda/src/descriptor/mod.rs#L684 to avoid the awkward `translate_pk2` APIs.

Reply:
`derived_descriptor()` is only available on `Descriptor<DescriptorPublicKey>`, so if I did a first `derive()` and now I have a `Descriptor<DerivedDescriptorKey>` then I can't use this directly and I have to go back to the "un-derived" descriptor.

Should I implement something similar for `Descriptor<DerivedDescriptorKey>`? Ideally with a different name, otherwise it can get confusing. Something that internally translates the keys